### PR TITLE
Update to clang-format-16

### PR DESCRIPTION
**Pull request by j-stephan**
It has been a while since we updated the clang-format style. This PR adds the rules added to clang-format since the last overhaul and removes those which are deprecated. I tried to follow the existing (manual) alpaka style where possible but we should vote on the changes.

### New rules (value after name = current proposed setting):
* AlignArrayOfStructures (`Left`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignarrayofstructures)
* BreakAfterAttributes (`Never`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#breakafterattributes)
* BreakBeforeConceptDeclarations (`Always`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#breakbeforeconceptdeclarations)
* BreakBeforeInlineASMColon (`Always`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#breakbeforeinlineasmcolon)
* EmptyLineAfterAccessModifier (`Never`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#emptylineafteraccessmodifier)
* EmptyLineBeforeAccessModifier (`Always`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#emptylinebeforeaccessmodifier)
* IndentAccessModifiers (`false`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#indentaccessmodifiers)
* IndentRequiresClause (`false`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#indentrequiresclause)
* InsertBraces (`true`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#insertbraces)
* InsertNewlineAtEOF (`true`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#insertnewlineateof)
* IntegerLiteralSeparator (`Binary: 4, Decimal: 3, Hex: 4`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#integerliteralseparator)
* LambdaBodyIndentation (`Signature`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#lambdabodyindentation)
* PenaltyBreakOpenParentheses (`2`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#penaltybreakopenparenthesis)
* PenaltyIndentedWhitespace (`2`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#penaltyindentedwhitespace)
* PPIndentWidth (`-1`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#ppindentwidth)
* QualifierAlignment (`Custom`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#qualifieralignment)
* QualifierOrder (`['friend', 'inline', 'static', 'constexpr', 'type', 'const', 'volatile', 'restrict']`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#qualifierorder)
* ReferenceAlignment (`Pointer`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#referencealignment)
* RemoveBracesLLVM (`false`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#removebracesllvm)
* RemoveSemicolon (`true`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#removesemicolon)
* RequiresClausePosition (`WithPreceding`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#requiresclauseposition)
* RequiresExpressionIndentation (`Keyword`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#requiresexpressionindentation)
* ShortNamespaceLines (`0`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#shortnamespacelines)
* SeparateDefinitionBlocks (`Always`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#separatedefinitionblocks)
* SpaceAroundPointerQualifiers (`Default`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#spacearoundpointerqualifiers)
* SpaceBeforeSquareBrackets (`false`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#spacebeforesquarebrackets)
* SpacesInLineCommentPrefix (`Minimum: 1, Maximum: -1`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#spacesinlinecommentprefix)

### Replacements:

* AlignTrailingComments (`Never`) - replaces old-style AlignTrailingComments (`false`) - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#aligntrailingcomments)
* LineEnding (`DeriveLF`) - replaces DeriveLineEnding - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#lineending)
* PackConstructorInitializers (`CurrentLine`) - replaces AllowAllConstructorInitializersOnNextLine and ConstructorInitializerAllOnOneLineOrOnePerLine - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#packconstructorinitializers)

### New project-specific options:

* AttributeMacros - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#attributemacros)
* IfMacros - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#ifmacros)
* WhitespaceSensitiveMacros - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#whitespacesensitivemacros)

### CODE REVIEW

**Comment by bernhardmgruber** [comment]
`BreakBeforeInlineASMColon`: I would go with `Never` or `OnlyMultiline` until we find a problem.
`InsertBraces`: I am super strongly against this one, because it just creates textual noise. Even worse, because alpaka already places each brace on a separate line.
`IntegerLiteralSeparator`: that's a fun one. I am curious how this will turn out.
`PenaltyBreakOpenParentheses` and `PenaltyIndentedWhitespace`: how do you come up with these numbers? why not just leave the defaults?
`PPIndentWidth`: again, just leave the default (which is -1).
`RemoveBracesLLVM`: as much as would like to see this as true, I know that older nvcc versions sometimes have parse errors when braces are missing on the else branch of an `if constexpr`.
`RequiresExpressionIndentation`: I am leaning towards `OuterScope` here

**Comment by j-stephan** [comment]
> `BreakBeforeInlineASMColon`: I would go with `Never` or `OnlyMultiline` until we find a problem.

I find inline ASM statements already hard to parse because we don't use them very often. Having them in multiple lines makes them easier to read IMO.

> `PenaltyBreakOpenParentheses` and `PenaltyIndentedWhitespace`: how do you come up with these numbers? why not just leave the defaults?

It isn't documented what the default is here. I just used the value from `PenaltyBreakAssignment` since that seams to work well.

>  `PPIndentWidth`: again, just leave the default (which is -1).

Just making the default explicit here.

**Comment by SimeonEhrig** [comment]
`InsertBraces`: I would prefer `true`. I had already the problem in the past, that I add a second line to a `if` body and forgot to add the braces. Have fun to find the strange behavior.

**Comment by SimeonEhrig** [comment]
But only, if you integrated it in your IDE. No one is running `clang-format` manually after code change.

**Comment by bernhardmgruber** [comment]
> But only, if you integrated it in your IDE. No one is running `clang-format` manually after code change.

I actually do both. CLion formats while a type, but sometimes I also do changes inside the diff viewer, or from the command line, in which case I run `git clang-format` :)

**Comment by j-stephan** [comment]
I'm a bit conflicted on the brace issue. I dislike it for this:

```c++
if(cond)
    one_liner_expression;
```

but would want it for this:

```c++
while (i--) 
  for (auto *A : D.attrs())
      handleAttr(A);
```

Unfortunately we cannot configure this beyond `true` or `false` so I'm slightly for inserting braces...

**Comment by bernhardmgruber** [comment]
@j-stephan given the two examples you listed, inserting braces will
> ```c++
> if(cond)
>     one_liner_expression;
> ```
double the screen space required for this piece of code. And for

> ```c++
> while (i--) 
>   for (auto *A : D.attrs())
>       handleAttr(A);
> ```
We add even more noise, since we turn 3 lines to 7. That is a crazy blow up of source code.

**Comment by psychocoderHPC** [comment]
I am against adding braces for single-line expressions. I know that there is the possibility to introduce errors during refactoring but adding braces is easily adding many lines per file.

**Comment by j-stephan** [comment]
Okay. So we are leaning towards `InsertBraces: false`. If there are no other wishes by the majority I'll change the line and then go ahead and merge this PR.

**Comment by psychocoderHPC** [comment]
`BreakBeforeInlineASMColon`: I am for OnlyMultiline
`IntegerLiteralSeparator` I do not like it or do not see why we need it.
`RequiresExpressionIndentation' I am for `OuterScope` because the other option looks like an alignment and we decided in the past to indent always and never align but option `KEyword` is aligning with the keyword.

**Comment by j-stephan** [comment]
I made the following changes:

* `AlignArrayOfStructures`: `None` - this is currently too buggy, especially in constructors
* `BreakBeforeInlineASMColon`: `OnlyMultiline`
* `InsertBraces`: `false`
* `IntegerDecimalSeparator`: added `DecimalMinDigits` and set it to `7`, otherwise this is too intrusive IMO
* `RequiresExpressionIndentation`: `OuterScope`
* `RemoveSemicolon`: `false` - detected false positives in alpaka which led to compilation failures
* `SortUsingDeclarations`: `Lexicographic` instead of `true` (I guess the latter was deprecated at some point)
* New project-specific options:
  * `ForEachMacros` - [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#foreachmacros)
  * `NamespaceMacros`: [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#namespacemacros)
  * `StatementMacros`: [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#statementmacros)
  * `TypenameMacros`: [Link](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#typenamemacros)

**Review by bernhardmgruber — APPROVED** [review]
